### DISCUSSION TITLE
Help wanted: pushing png to sprite

### DISCUSSION THREAD

**Discussion by monte-monte** [Bodmer/TFT_eSPI#2418] [Q&A - Sprites]
I've adapted example code to show png file from LittleFS on M5Stack core. Now I want to push said image to sprite for further manipulations, but it doesn't work, sprite ends up filled with lines like a bar code (sorta). First I tried ``pushMaskedImage`` that hanged completely, now I simply try to ``pushImage`` to sprite, but something seems wrong. Tried creating smaller sprite, result is the same.

IDE PlatformIO, TFT_eSPI v.2.5.21

**Comment by Bodmer** [comment]
The pushMaskedImage() function can only be used to push images to the TFT.

PNG images can be pushed to a sprite and this works for me so I assume there is an error in your sketch.

**Comment by monte-monte** [comment]
@Bodmer this is a code I am using:
```
TFT_eSprite sprite = TFT_eSprite(&lcd);

void * pngOpen(const char *filename, int32_t *size) {
  Serial.printf("Attempting to open %s\n", filename);
  pngfile = LittleFS.open(filename, "r");
  *size = pngfile.size();
  return &pngfile;
}

void pngClose(void *handle) {
  File pngfile = *((File*)handle);
  if (pngfile) pngfile.close();
}

int32_t pngRead(PNGFILE *page, uint8_t *buffer, int32_t length) {
  if (!pngfile) return 0;
  page = page; // Avoid warning
  return pngfile.read(buffer, length);
}

int32_t pngSeek(PNGFILE *page, int32_t position) {
  if (!pngfile) return 0;
  page = page; // Avoid warning
  return pngfile.seek(position);
}

void pngDraw(PNGDRAW *pDraw) {
  uint16_t lineBuffer[MAX_IMAGE_WDITH];
  png.getLineAsRGB565(pDraw, lineBuffer, PNG_RGB565_BIG_ENDIAN, 0xffffffff);
  sprite.pushImage(0, 0 + pDraw->y, pDraw->iWidth, 1, lineBuffer);
}

void drawSplash() {
    sprite.createSprite(125, 125);
    int16_t rc = png.open("/srl_logo.png", pngOpen, pngClose, pngRead, pngSeek, pngDraw);
    if (rc == PNG_SUCCESS) {
      Serial.printf("image specs: (%d x %d), %d bpp, pixel type: %d\n", png.getWidth(), png.getHeight(), png.getBpp(), png.getPixelType());
      if (png.getWidth() > MAX_IMAGE_WDITH) {
        Serial.println("Image too wide for allocated line buffer size!");
      }
      else {
        rc = png.decode(NULL, 0);
        png.close();
      }
      sprite.pushSprite(98,58);
    }
    sprite.deleteSprite();
}
```
Also tried to change color depth of a sprite to 1 bit, didn't help.
Any other usage of sprite has no issues.
Pushing this same image to the display directly has no issues either.

This is how image looks with this code: 
![image](https://user-images.githubusercontent.com/6649967/219962348-59a08fd5-27a4-4a34-b778-5e2d0813d92c.png)


And this is pushing to the display:
![image](https://user-images.githubusercontent.com/6649967/219962310-b73579e4-357c-412e-8bfa-14247fc68aae.png)

**Comment by Bodmer** [comment]
I have created the example below and it is working fine. I suspect the sketch in your application is much more complex and that the memory available is insufficient for the PNGDEC library to run correctly.
[LittleFS_PNG_2418.zip](https://github.com/Bodmer/TFT_eSPI/files/10778140/LittleFS_PNG_2418.zip)

**Comment by monte-monte** [comment]
Okay, thanks for the help. I figured out that the culprit was the function that used the same sprite object and was called before ``drawSplash()``. It sets color depth to 1 bit, and I assumed for some reason that this setting is reset on ``deleteSprite()``. So it tried to push 16 bit data to 1 bit sprite, and those stripes were the result of it.
Strange behavior, but I'm glad that we were able to figure this out. 
By the way, there is plenty of memory left while using this sprite with PNGdec, ``ESP.getFreeHeap()`` returns 247372.
